src/fs/filter.rs

```diff
@@ -94,6 +103,52 @@ impl FileFilter {
         }
     }
 
+    /// remove directories if they contain a CACHEDIR.TAG with the correct signature.
+    /// does nothing if `self.cachedir_ignore == Off`
+    pub fn filter_cachedirs(&self, files: &mut Vec<File<'_>>) {
+        if self.ignore_cachedir == IgnoreCacheDir::CheckAndIgnore {
+            files.retain(|f| {
+                if !f.is_directory() {
+                    return true;
+                }
+                let Ok(read_dir) = std::fs::read_dir(&f.path) else {
+                    return true;
+                };
+                let mut found_tag = false;
+                for child in read_dir {
+                    let Ok(child) = child else {
+                        continue;
+                    };
+                    let is_cachedir_tag = Self::is_cachedir_tag(&child.path());
+                    if is_cachedir_tag {
+                        found_tag = true;
+                        break;
+                    }
+                }
+                !found_tag
+            });
+        }
+    }
+
+    /// check if `path` is named "CACHEDIR.TAG" and has the correct magic number ([`CACHEDIR_MAGIC`]).
+    fn is_cachedir_tag(path: &std::path::PathBuf) -> bool {
```

Comment:
Can we have this accept a `&std::path::Path`? I believe its comparable to accepting a `&str` over `&String` where it will still allocate when not needed.

Reply:
done (oops my squashing outdated this)